internal/brokerservice/brokerservice.go

```diff
@@ -107,7 +108,12 @@ func (s *Service) Serve() error {
 
 // Stop stop the service and do all the necessary cleanup operation.
 func (s *Service) Stop() error {
-	close(s.serve)
-	s.disconnect()
+	// Check if already stopped.
+	select {
+	case <-s.serve:
+	case <-time.After(10 * time.Millisecond):
+		close(s.serve)
```

Comment:
unsure how this fixes the race/double close case? was it something we were doing in adsys or authd?

Reply:
Receiving from a closed channel is a non-blocking no-op. If the channel is not closed, it will hang there the same way as the `serve()` function does and then we close it after a little delay.

TLDR: If the channel is already closed, Stop() is a no-op.

Reply:
instead of a small delay (which introduce a bigger race window if we don’t use mutex), why not use `default:`?

Reply:
You are very smart, mister. I didn't think of it at all 🤣